Maple2.Server.World/Containers/PartyManager.cs

```diff
@@ -0,0 +1,216 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Grpc.Core;
+using Maple2.Model.Error;
+using Maple2.Model.Game;
+using Maple2.Server.Channel.Service;
+using ChannelClient = Maple2.Server.Channel.Service.Channel.ChannelClient;
+
+namespace Maple2.Server.World.Containers;
+
+public class PartyManager : IDisposable {
+    public required ChannelClientLookup ChannelClients { get; init; }
+    public readonly Party Party;
+    private readonly ConcurrentDictionary<long, (string, DateTime)> pendingInvites;
+
+    public PartyManager(Party party) {
+        Party = party;
+        pendingInvites = new ConcurrentDictionary<long, (string, DateTime)>();
+    }
+
+    public void Dispose() { }
+
+    public void Broadcast(PartyRequest request) {
+        if (request.PartyId > 0 && request.PartyId != Party.Id) {
+            throw new InvalidOperationException($"Broadcasting {request.PartyCase} for incorrect party: {request.PartyId} => {Party.Id}");
+        }
+
+        request.PartyId = Party.Id;
+        foreach (IGrouping<short, PartyMember> group in Party.Members.Values.GroupBy(member => member.Info.Channel)) {
+            if (!ChannelClients.TryGetClient(group.Key, out ChannelClient? client)) {
+                continue;
+            }
+
+            request.ReceiverIds.Clear();
+            request.ReceiverIds.AddRange(group.Select(member => member.Info.CharacterId));
+
+            try {
+                client.Party(request);
+            } catch { }
+        }
+    }
+
+    public void FindNewLeader(long characterId) {
+        PartyMember? newLeader = Party.Members.Values.FirstOrDefault(m => m.CharacterId != characterId);
+        if (newLeader == null) {
```

Comment:
Maybe disband here? I can't think of an instance where the party should still exist if this happens.